docs/DESIGN.md

```diff
@@ -0,0 +1,77 @@
+# YesNo Design
+
+This document is meant to describe the guiding principles and direction of yesno development. When adding new features or prioritizing the backlog, these ideas should help in decision making.
+
+## Usage Types
+
+1. `yesno.save, yesno.load`
+1. `yesno.record, yesno.complete`
+1. `yesno.test`
+
+Case 1 is the most basic and requires the most work by the user. It is intended as an escape hatch for users that have unique or specific needs that the default usage doesn't meet. Most users should be using cases 2 and 3 (where case 3 is just a convience testing wrapper around case 2).
+
+## mockRule Design
+
+A new mockRule method will be added to allow the user to define a set of rules that will be evaluated in the order that they are defined. This will let the user control how yesno will handle recording and responding. By default, yesno will record all network requests and mock them in the order in which they were recorded. Using mockRule, matching requests can be made live instead of mocked, or the mock response can be specified instead of using the recorded value. As soon as a rule matches a request, the rule is applied and processing stops. If the user wants the default action to be live requests, then they would just create a last rule that matches everything to be live.
+
+## YesNo Interfaces
+
+- `yesno.mockRule` - set of user defined rules to be applied to all requests
+  - `yesno.mockRule().record()` - matching requests will be recorded/mocked
+  - `yesno.mockRule().live()` - matching requests will be proxied
+  - `yesno.mockRule().respond()` - matching requests will be mocked with the defined response
+  - `yesno.mockRule().redact()` - matching properties will be redacted
+- `yesno.matching()` - used after responses have been intercepted to filter the collection
+- `yesno.mockMatchingFunction()` - allows the user to override the default matching function, the callback gets passed:
+  - the request data object
+  - the collection of mocks
+  - returns true on match
+
+## Example Use Cases
+
+Record and mock all requests (default use case)
```

Comment:
The first two examples would need strict vs not strict mode right? I don't see that concept described at all.

Reply:
I was thinking that we probably don't need to specify a `strict` mode if we allow the user to set the number of times a mock can be used/matched.  We can default it to 1 for the strict normal behavior and a user could set it to a max number or 0 for unlimited matching.  Something like:
`yesno.mockRule(/A/).record().maxMockRequests(5)`
Then it would use the mock for A 5 times before it is considered "used"